File: app/src/main/java/ua/com/foxminded/locationtrackera/model/auth/FirebaseAuthNetwork.java
package ua.com.foxminded.locationtrackera.model.auth;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.concurrent.ExecutionException;

import io.reactivex.rxjava3.core.Single;
import ua.com.foxminded.locationtrackera.util.Result;

public class FirebaseAuthNetwork implements AuthNetwork {

    private final FirebaseAuth firebaseAuth;

    public FirebaseAuthNetwork(FirebaseAuth firebaseAuth) {
        this.firebaseAuth = firebaseAuth;
    }

    @Override
    public Single<Result<Void>> firebaseRegister(String username, String email, String password) {
        return Single.fromCallable(() -> {
            final Task<AuthResult> task = firebaseAuth.createUserWithEmailAndPassword(
                    email, password
            );

            try {
                Tasks.await(task);
            } catch (ExecutionException e) {
                e.printStackTrace();
            }

            if (task.isSuccessful()) {
                final Task<Void> task2 = FirebaseFirestore.getInstance()
                        .collection(FirebaseNetworkConstants.COLLECTION_PATH_USERS)
                        .document(firebaseAuth.getCurrentUser().getUid())
                        .set(new User(username, email));

                try {
                    Tasks.await(task2);
                } catch (ExecutionException e) {
                    e.printStackTrace();
                }

                if (task.isSuccessful()) {
                    return new Result.Success<>(null);
                } else {
                    return new Result.Error(task2.getException());
                }
            } else {
                return new Result.Error(task.getException());
            }
        });
    }

    @Override
    public Single<Result<Void>> firebaseLogin(String email, String password) {
        return Single.fromCallable(() -> {
            final Task<AuthResult> task = firebaseAuth.signInWithEmailAndPassword(email, password);

            try {
                Tasks.await(task);
            } catch (ExecutionException e) {
                e.printStackTrace();
            }

            if (task.isSuccessful()) {
                return new Result.Success<>(null);
            } else {
                return new Result.Error(task.getException());
            }
        });
    }

    @Override
    public Single<Result<Void>> resetPassword(String email) {
        return Single.fromCallable(() -> {
            final Task<Void> task = firebaseAuth.sendPasswordResetEmail(email);

            try {
                Tasks.await(task);
            } catch (ExecutionException e) {
                e.printStackTrace();
            }

            if (task.isSuccessful()) {
                return new Result.Success<>(null);
            } else {
                return new Result.Error(task.getException());
            }
        });
    }

    @Override
    public void logout() {
        firebaseAuth.signOut();
    }
}

File: models_impl/src/main/java/ua/com/foxminded/locationtrackera/models_impl/auth/FirebaseAuthNetwork.kt
package ua.com.foxminded.locationtrackera.models_impl.auth

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await
import ua.com.foxminded.locationtrackera.models.auth.AuthNetwork
import ua.com.foxminded.locationtrackera.models.auth.User
import ua.com.foxminded.locationtrackera.models.util.Result

class FirebaseAuthNetwork(private val firebaseAuth: FirebaseAuth) : AuthNetwork {

    override suspend fun firebaseRegister(
        username: String, email: String, password: String
    ): Result<Void?> {
        return try {
            firebaseAuth.createUserWithEmailAndPassword(email, password).await()
            try {
                FirebaseFirestore.getInstance()
                    .collection(COLLECTION_PATH_USERS)
                    .document(firebaseAuth.currentUser!!.uid)
                    .set(User(username, email))
                    .await()
                Result.Success(null)
            } catch (e: Exception) {
                Result.Error(e)
            }
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    override suspend fun firebaseLogin(email: String, password: String): Result<Void?> {
        return try {
            firebaseAuth.signInWithEmailAndPassword(email, password).await()
            Result.Success(null)
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    override suspend fun resetPassword(email: String): Result<Void?> {
        return try {
            firebaseAuth.sendPasswordResetEmail(email).await()
            Result.Success(null)
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    override fun logout() {
        firebaseAuth.signOut()
    }

    override fun isUserLoggedIn() = firebaseAuth.currentUser != null

    companion object {
        private const val COLLECTION_PATH_USERS = "Users"
    }
}
